object.c

```diff
@@ -120,7 +120,7 @@ rb_obj_reveal(VALUE obj, VALUE klass)
 VALUE
 rb_obj_setup(VALUE obj, VALUE klass, VALUE type)
 {
-    VALUE ignored_flags = RUBY_FL_PROMOTED;
+    VALUE ignored_flags = RUBY_FL_PROMOTED | RUBY_FL_SEEN_OBJ_ID;
```

Comment:
I think it would be valuable to define some mask such as `RUBY_FL_INHERITED_ON_CLONE` in `fl_type.h` to make it easier to reason about. But since this is meant as a bugfix, I'd rather make the change as small as possible to make it easier to backport.

Also this line changed in 3.3, so the backport to 3.2 and 3.1 will conflict, but it should be easy to resolve.